doc/developer/design/20231216_retain_definite_error_status.md

```diff
@@ -0,0 +1,49 @@
+# Retaining definite status errors across restart
+
+## The Problem
+
+https://github.com/MaterializeInc/materialize/issues/23950
+
+There are a number of annoying deficiencies in how we report statuses for
+sources and sinks, the worst of which is that we do not currently ensure that a
+source or a sink that encountered a definite error always has that error in its
+entry in the statuses table.
+
+For example, if you perform an incompatible alter for a PG subsource and then
+restart MZ, the status will show that the subsource is either starting or
+running, while reading it will always return a definite error.
+
+## Success Criteria
+
+Restarting MZ retains definite errors in the status table for the source or
+sink.
+
+## Out of Scope
+
+Imagining a world where we make definite errors retractable.
+
+## Solution Proposal
+
+We can solve this issue by introducing a new status, which I'll call "ceased."
```

Comment:
This is a fine solution but it does paper over a more fundamental issue that once fixed will render the need for a special "ceased" status obsolete. The problem lies in the function signature of `SourceRender::render` which is required to return a *single* collection of type `Collection<G, (usize, ...)>`. The `usize` in there is used to later break apart the stream into one per subsource, but that has the downside that there is a single shared progress reported. This means that the progress reported must necessarily be the minimum across all subsources, even if particular subsources are far ahead in the future (e.g because their snapshot has finished or because they hit a permanent error like an incompatible schema change). 

The solution would be if `SourceRender::render` returned a `Vec<Collection<G, ...>>` so that each subsource's progress is tracked separately. Then, the source implementation would be able to advance individual subsources to the empty frontier indicating finality. Status reporting would then just piggy back on that and know whether a subsource is done or not, and would not re-report a starting state.

It would be amazing if we could use this effort to improve the API of this fundamental trait. What do you think?